dotCMS/src/main/java/com/dotcms/csspreproc/dartsass/DartSassCompiler.java

```diff
@@ -0,0 +1,230 @@
+     * execution of the command.
+     */
+    public RuntimeUtils.TerminalOutput terminalOutput() {
+        return this.terminalOutput;
+    }
+
+    /**
+     * Compiles the incoming SCSS File and produces the expected CSS File.
+     * <p>After validating the involved CSS Files and the location of the Dart SASS library, the final compilation
+     * command is put together based on the specified parameters. Executing it via the Java API will result in the
+     * expected CSS File that will be returned to the browser.</p>
+     *
+     * @return An {@link Optional} object containing the results of the compilation.
+     *
+     * @throws DotSassCompilerException An error occurred when compiling the SCSS File.
+     */
+    public Optional<String> compile() throws DotSassCompilerException {
+        validateInputData();
+        final String[] compileInstructions = buildCompilationCommand();
+        this.terminalOutput = RuntimeUtils.runProcessAndGetOutputWithError(compileInstructions);
+        try {
+            return FileUtil.read(this.outputFile());
+        } catch (final IOException e) {
+            final String errorMsg = String.format("An error occurred when reading resulting CSS file '%s': %s",
+                    this.outputFile().getAbsolutePath(), e.getMessage());
+            Logger.error(this, errorMsg, e);
+            throw new DotSassCompilerException(errorMsg, e);
+        }
+    }
+
+    /**
+     * Generates the CLI call to the Dart SASS compiler based on the location of the library in dotCMS, and the
+     * specified configuration parameters that define the behavior of the compiler.
+     *
+     * @return An array containing the parts that make up the compile command.
+     */
+    private String[] buildCompilationCommand() {
+        final List<String> compileInstructions = new ArrayList<>();
+        final String location = dartSassLocation.get().endsWith(File.separator) ? dartSassLocation.get() :
+                                        dartSassLocation.get() + File.separator;
+        compileInstructions.add(location + getOsSpecificSassCommand());
+        compileInstructions.addAll(this.compilerOptions().generate());
+        compileInstructions.add(this.inputFile().getAbsolutePath());
+        compileInstructions.add(this.outputFile().getAbsolutePath());
+        return compileInstructions.toArray(String[]::new);
+    }
+
+    /**
+     * Utility method used to validate that the information required by the compiler is available and valid right before
+     * carrying out the compilation process.
+     *
+     * @throws DotSassCompilerException An error occurred when validating required compiler input data.
+     */
+    private void validateInputData() throws DotSassCompilerException {
+        if (null == this.inputFile() || !this.inputFile().exists()) {
+            throw new DotSassCompilerException("Input SCSS file is null or doesn't exist");
+        }
+        if (null == this.outputFile()) {
```

Comment:
can add if canWrite

Reply:
The resulting CSS file does not exist in FS at that point yet. Checking write permissions throws an exception.